### ISSUE TITLE
Numpy bfloat16 support

### ISSUE THREAD

**Issue by jakpiase** [numpy/numpy#19808]
## Feature
Hi,
I am working at PaddlePaddle(chinese DL framework). We and other DL frameworks would extremely benefit from integrated bfloat16 numpy datatype. I have seen that TF added its own implementation and lately a standalone pip package [bfloat16](https://pypi.org/project/bfloat16/) was released. 
I have also seen NEP 41, 42, 43 which from what I understand will allow adding new, user-defined datatypes. Are you guys planning to integrate bfloat16 into core numpy? If you don't have the bandwidth to do that, is there anyone that can guide me how to implement that so I can make a PR? And how are these NEPs going?

**Comment by seberg** [comment]
NEPs 41-43 are mostly implemented (there is one major step remaining, that is mostly finished though; there will be few smaller steps until we are there though, and I expect things will come up as soon as someone tries to write a more complex dtype – which bfloat16 is probably not though).

However, it is not yet public API, I hope to followup with an experimental public API very soon, though.  And there is a repo (which has to change a bit soon) [here](https://github.com/seberg/experimental_user_dtypes).

I am personally not sure whether or not we should include `bfloat16` in NumPy proper, there are some advantages, and some disadvantages.  (NumPy dtypes are a bit strange sometimes, defining `bfloat16` at least somewhat outside of NumPy would sidestep this issue, by being very clear about it being not one of the "main" NumPy dtypes.)

But, opinions will probably differ, I would somewhat prefer a `numpy_bfloat16` package right now, the main problem is that not including it NumPy means it needs the new DType API to be out of an experimental state probably (or maybe strict NumPy version binding. I.e. force the user to install `numpy_bfloat16` with a version of NumPy that it knows it will work with correctly.)

Or, we include it in NumPy (for ABI reasons) but make it accessible through its own import?

**Comment by leofang** [comment]
> Or, we include it in NumPy (for ABI reasons) but make it accessible through its own import?

This would be my preference for bfloat16 (the current issue) and complex32 (https://github.com/numpy/numpy/issues/14753). Would be very useful for GPU libraries like CuPy (and especially CuPy, which does not have its own type system and simply uses the dtypes from NumPy).

**Comment by seberg** [comment]
No matter the outcome of the discussion, I think it may make sense to start with a new `bfloat16` pypi package as soon as the experimental API for it exists (which I can expedite a bit more if someone wants to work on creating such a prototype package – its great for motivation ☺; and it already exists enough that one could start IMO).

The reason is that I think there are a few questions that need answering, e.g. whether we even want a `bfloat16` scalar.

**Comment by jakirkham** [comment]
Agreed. Just to add to that, there may be issues in a `bfloat16` implementation that may need faster iteration on to address. Having it live outside NumPy for a while until things mature is likely good both for the `bfloat16` implementation and NumPy.

Sebastian, do you have thoughts on whether having a new repo live here (under the NumPy org) would make sense? Or is there some incubator or contrib NumPy org that would make sense? How should experimental things like this develop?

**Comment by alvarosg** [comment]
+1 to adding this support. 

From the thread in this [JAX bug](https://github.com/google/jax/issues/11014), it seems that there are several inconsistencies in type promotion behavior for the current custom `bfloat16` type used by both TensorFlow and JAX, which cannot be fixed without actually adding better native support in numpy.

**Comment by seberg** [comment]
> which cannot be fixed without actually adding better native support in numpy

This is not true anymore, they *can* be addressed in a fully compatible way with the new (although not yet public) API.  Yes, there will be road-bumps, but the point is that there is a path of starting this outside of NumPy and _then_ considering whether that is good or whether inclusion is desired.

**Comment by hawkinsp** [comment]
I wrote that bfloat16 extension (the TF/JAX one) targeting NumPy 1.16 when working on early TPUs. NumPy's type extension APIs have certainly advanced since then. So it's probably time someone had a go at adapting it to the new APIs.

In passing: one other observation I will make while I am here is that at least two varieties of 8-bit floating point type are gaining popularity in the machine learning world as well (e.g., non-standard E4M3 and E5M2 types from hardware vendors), and we were looking at generalizing the extension to support those as well.

**Comment by alvarosg** [comment]
(@jakevdp for visibility)

Thanks! In that case, @hawkinsp would it be useful if I start a bug similar to google/jax#11014 , but in the tensorflow repo?

**Comment by hawkinsp** [comment]
The JAX repo bug is fine: in practice we are the ones working on this and that's the project we work in most.

(Really this extension should be its own pip package, and we may do that. I note that someone did fork our code and do this already (`bfloat16`), so I think there are at least 4 packages through which you can get what is in essence the same extension now, and that's a mess that needs to be cleaned up.)

**Comment by jakpiase** [comment]
Hi @hawkinsp, I am an author of [paddle_bfloat](https://pypi.org/project/paddle-bfloat) package which is an upgraded version of "original" [bfloat16](https://pypi.org/project/bfloat16) package which contains some minor fixes(fixing compilator warnings, some missing functions on windows etc). If I could help you in any way, please let me know.

**Comment by mattip** [comment]
@BlueskyFR We will adopt this approach
> ... there is a path of starting this outside of NumPy and then considering whether that is good or whether inclusion is desired.

so please help one of the implementations mentioned here to reach maturity.

**Comment by BlueskyFR** [comment]
@mattip I don't really understand what is going on, why is it developed as external repos and not as a Numpy branch?

**Comment by mattip** [comment]
NumPy is very conservative when it comes to enhancements since any new NumPy enhancement becomes part of the entire scientific python stack. There are too many unknowns with `bfloat16` support by compilers and hardware for it to be immediately part of NumPy, the team would prefer the code mature outside of NumPy first. 

The Data API has also not adopted `bfloat16`, see point 4 on what is [out of scope](https://data-apis.org/array-api/latest/purpose_and_scope.html?highlight=bfloat16#out-of-scope) for the standard.

The approach to first have code as a stand-alone library was successfully used when revamping the `numpy.random` module: it was first developed as part of [randomgen](https://github.com/bashtage/randomgen), then [NEP 19](https://numpy.org/neps/nep-0019-rng-policy.html) was written about the migration path, and only then the code was merged to NumPy.

**Comment by BlueskyFR** [comment]
Okay, thanks for the explanation. Still, developing a feature as a branch of the repo doesn't mean that it is **released** with it, it would just be a common place to develop such a feature. The problem with waiting for external libraries to do that is that community effort is extremely limited whereas the Numpy repo can bring people together.

I am affraid that waiting for some external support requires a couple more years before reaching something stable, so IMO the Numpy repo has the capacity to help this effort with endangering the public stable releases.

**Comment by seberg** [comment]
@BlueskyFR doing a branch in NumPy means you get slowed down by PRs to the NumPy repo.  Slower review, etc.  If there was a champion who is already a NumPy dev there may be a point to it, but we do not have that.

We can still consider putting it into the NumPy organization, but I am not sure that it would help much.  NumPy does not have resources to push this, having a feature branch that doesn't move quickly will not be helpful IMO.

You can use this issue, the NumPy mailing list, and https://discuss.scientific-python.org to coordinate.  I can try to keep an eye on things also to have a NumPy dev in the loop, but we are not the ones that can drive this for now.

**Comment by BlueskyFR** [comment]
> @BlueskyFR doing a branch in NumPy means you get slowed down by PRs to the NumPy repo.  Slower review, etc.  If there was a champion who is already a NumPy dev there may be a point to it, but we do not have that.
> 
> We can still consider putting it into the NumPy organization, but I am not sure that it would help much.  NumPy does not have resources to push this, having a feature branch that doesn't move quickly will not be helpful IMO.
> 
> You can use this issue, the NumPy mailing list, and https://discuss.scientific-python.org to coordinate.  I can try to keep an eye on things also to have a NumPy dev in the loop, but we are not the ones that can drive this for now.

Okay, many thanks for the replies!

**Comment by ogrisel** [comment]
> The problem with waiting for external libraries to do that is that community effort is extremely limited whereas the Numpy repo can bring people together.

@BlueskyFR if you start such an external implementation, please put the a link to it in a comment here (or even in the description of this issue) so that people googling for "numpy bfloat16" and end up here as I just did will find there way to your project and start consolidating efforts there.

**Comment by jakevdp** [comment]
Hi all – we just released `ml_dtypes`: https://github.com/jax-ml/ml_dtypes: it's a lightweight stand-alone implementation of bfloat16 and a few other dtypes used in machine learning contexts, extracted from the source code of JAX/XLA. We plan to maintain this actively going forward as the source for dtype definitions in JAX and other libraries. The code is permissively licensed, with the goal that if NumPy were ever to add a native bfloat16 dtype, it could adapt it from the implementation in `ml_dtypes`.

**Comment by loretoparisi** [comment]
Is the solution of casting to half precision https://github.com/pytorch/pytorch/issues/90574
```python
if embeddings.dtype == torch.bfloat16:
    embeddings = embeddings.float()
embeddings = embeddings.numpy()
```
proper for `nn.Embedding` layers trained with mixed precision:

```python
trainer = L.Trainer(
        logger=logger,
        log_every_n_steps=1,
        # '16-true', '16-mixed', 'bf16-true', 'bf16-mixed', '32-true', '64-true', 64, 32, 16, '64', '32', '16', 'bf16')
        precision="bf16-true",
        callbacks=[checkpoint_callback],
        accelerator=accelerator,
        devices=devices,
        num_nodes=num_nodes,
        strategy=strategy,
        limit_train_batches=1.0,
        max_epochs=hparams["epochs"],
        deterministic=True
    )
```
Noteworthy, a train strategy also need the same `dtype`:

```python
strategy = FSDPStrategy(
                cpu_offload=True,
                activation_checkpointing_policy=encoder_decoder_policy,
                auto_wrap_policy=embedding_auto_wrap_policy,
                # for numerical stability mixed precision keeps the model weights in full float32 precision 
                # while casting only supported operations to lower bit precision.
                mixed_precision=MixedPrecision(param_dtype=torch.torch.bfloat16, cast_forward_inputs=True),
                process_group_backend="nccl",
                sharding_strategy="FULL_SHARD"
            )
```

**Comment by rgommers** [comment]
There were five separate feature requests for `bfloat16` including this one; I have now closed the other four as duplicates. 

There does seem to be a high level of interest in this feature. @GaelVaroquaux you were also mentioning this as being desirable for scikit-learn recently - maybe you or @ogrisel can write a summary here of what native `bfloat16` support in NumPy would bring for you (in general, and also why `ml_dtypes` isn't enough)?

@leofang same question for you from the CuPy perspective, now that that is moving forward (xref [this section of the CuPy v14 relnotes](https://docs.cupy.dev/en/stable/upgrade.html#minimal-support-for-bfloat16)).

**Comment by hawkinsp** [comment]
Speaking as one of the ml_dtypes authors:
* we would be thrilled for upstream NumPy to add any of the types that we have. `bfloat16` is the most clearly useful candidate, and supported on many kinds of hardware at this point. We may or may not have time to actually do this but it would have our full support.
* there is a zoo of other float types around, and it may be too early for NumPy to add most of them, even if it ever does. So `ml_dtypes` still plays an important role.
* if I had one ask from NumPy, it would be that we resolve the remaining issues that make it hard for user-defined dtypes to act as first class `float` or `complex` dtypes. That would reduce some of the pressure to upstream. @seberg was already looking at this.

**Comment by ogrisel** [comment]
I suppose that several machine learning models in scikit-learn do not really need high precision floating point arithmetics and could run more memory efficiently with `bfloat16` either to store data or model parameters (or both).

That being said, making sure that code works reasonably well with low precision arrays would need careful work to identify where precision is needed or not to upcast to `float32` as needed (e.g. to accumulate values in a loop) and then downcast back to `bloaft16` when appropriate to save memory.

I think the following example from `ml_dtypes` explain why it would be challenging to leverage `bfloat16` in scikit-learn: one cannot expect to naively write regular numpy code and expect it to work out of the box as it's really quick to hit precision limits:

https://github.com/jax-ml/ml_dtypes?tab=readme-ov-file#quirks-of-low-precision-arithmetic

Furthermore, to really benefit from `bfloat16` in scikit-learn, we would probably need access to a `bfloat16` aware implementation of BLAS level 1 to 3. I don't know what is the status `bfloat16` support in OpenBLAS and which hardware it optimizes for. At the moment the numpy would transparently upcast to float32 to perform BLAS operations:

```python
>>> import ml_dtypes, numpy as np
>>> A = np.random.randn(1024, 1024).astype(ml_dtypes.bfloat16)
>>> A.dtype
dtype(bfloat16)
>>> (A @ A.T).dtype
dtype('float32')
```

**Comment by GaelVaroquaux** [comment]
To my eyes, the greatest benefit of having bfloat16 would be to be able to apply some of scikit-learn's functions on data in deep-learning pipelines without type conversion and ideally with less memory copies.

Some of these functions could be very simple ones, such as one-hot encoding or standard scaling. I don't even have in mind linear algebra.

**Comment by rgommers** [comment]
> If I had one ask from NumPy, it would be that we resolve the remaining issues that make it hard for user-defined dtypes to act as first class `float` or `complex` dtypes. That would reduce some of the pressure to upstream. [@seberg](https://github.com/seberg) was already looking at this.

Makes sense. I found [this write-up](https://hackmd.io/@icU94i8ySRKSbk8eiRWndw/Bk4T45uWn) from @jakevdp again, which says:

- No way to register new dtypes with `numpy.finfo`
- No way to make this a true subtype of `numpy.floating` because of hard-coded logic about float widths
- Saving as `*.npy` does not work because there is no unique char code for bfloat16

Having a regular name in the main namespace is of course also useful.



> That being said, making sure that code works reasonably well with low precision arrays would need careful work to identify where precision is needed or not to upcast to `float32` as needed (e.g. to accumulate values in a loop) and then downcast back to `bloaft16` when appropriate to save memory.

I'd expect `bfloat16` to be implemented as a "storage type". Meaning all calculations will be done in `float32`, only manipulations (things like `concat`, `moveaxis`, comparisons, etc.) don't need such a conversion. That's how `float16` is implemented as well. And on CPU that's what PyTorch and JAX do too - AFAIK there isn't a single library that does everything natively in `bfloat16` on CPU. There is no hardware support, nor `libm` support. Both `float16` and `bfloat16` tend to be stored as `uint16_t`, and to do anything nontrivial it needs conversion to `float32` (there are a few exceptions, like dedicated AVX512 instructions that can be leveraged on x86-64; but not many). 

The benefits come from lower memory usage, needing only half the memory bandwidth (which is often the bottleneck), and usefulness when writing device-agnostic code.

**Comment by leofang** [comment]
> [@leofang](https://github.com/leofang) same question for you from the CuPy perspective, now that that is moving forward (xref [this section of the CuPy v14 relnotes](https://docs.cupy.dev/en/stable/upgrade.html#minimal-support-for-bfloat16)).

bf16 is one of the most requested dtypes in CuPy, both internally in NVIDIA and externally (ex: https://github.com/cupy/cupy/issues/7527). 

My understanding for the latter is that users want native support eventually (ex: arithmetic ops, ufuncs, reductions, linalg, ...), but the first step is to make a type identifier so that we can speak in the same language with the same vocabulary, and for that purpose what's offered by `ml_dtypes` is more than enough; we can handle all the lowering and promotion rules internally in CuPy ourselves, we just need a type identifier that works for both NumPy/CuPy. (@seberg please keep me honest and accurate! 🙂)

For the former, once we have the vocabulary type we can enable kernel development across host/device boundary, see e.g. [a working example for NumPy + CuPy + Numba-cuda + cuda-core + ml_dtypes](https://github.com/NVIDIA/numba-cuda/blame/893dde7ed75ade4ef44b42ed0163e53982fbcf49/numba_cuda/numba/cuda/tests/cudapy/test_cupy_bfloat16_interop.py#L30-L47).

**Comment by seberg** [comment]
CuPy implements the ufunc promotion part (it is also just simpler) and also relies on `result_type`, scalars, etc. (this is no problem).

CuPy may inherit NumPy limitations in some parts, although I don't think it is much (besides `np.save`).  E.g. it's matrix multiply already uses float16/bfloat16 (with float32 as precision type).

There are two advantages in just including it (I am not opposed). First, it is more convenient as use increases and is clearly available as `np.bfloat16` (this is the Array API user argument, unless we say `array-api-compat` or so is enough). Second, we have to start playing whack-a-mole with issues.  Integrating to NumPy is probably quite a bit of work, but the long tail of whack-a-mole may be easier.
Maybe there isn't any big downside? Part of me likes an odd-one-out type outside NumPy to manage the expectation that it may behave odd occasionally. And I am not sure we can get away with just declaring it a storage type before long (we already have a push for float16 SIMD loops although we have to discuss reverting it!). Of course, one could argue that doing that outside of NumPy isn't much nicer.

**Comment by hpkfft** [comment]
For what it's worth, I did initiate a discussion for adding [bfloat16 support to struct](https://discuss.python.org/t/add-bfloat16-support-to-struct/106723) in Python.  The response was generally negative (to my mind), partly because NumPy does not support bfloat16.

Regarding `ml_dtypes`, I would like to mention the following interoperability issue:
```
>>> import numpy as np
>>> import ml_dtypes
>>> a = np.array([3.25, 7.5], dtype=ml_dtypes.bfloat16)
>>> a
array([3.25, 7.5], dtype=bfloat16)
>>> a.__dlpack__()
Traceback (most recent call last):
  File "<python-input-4>", line 1, in <module>
    a.__dlpack__()
    ~~~~~~~~~~~~^^
BufferError: DLPack only supports signed/unsigned integers, float and complex dtypes.
```
Note that DLPack does have enumerator `kDLBfloat` for bfloat16, but NumPy does not recognize `ml_dtypes.bfloat16`.

**Comment by seberg** [comment]
FWIW, I would be happy if we can come up with a quick solution for that and it may not have to be perfect.  First, we might even be able to do that for NumPy 2.5 already and second, you really want support for all the other narrow dtypes, too.  And I doubt those will make the cut.

We need a bit of input/discussion of what is the right approach.  Even something like `np.dtypes.register_to_dlpack_map(int ID, dtype)` may be OK (would probably update a dict, maybe warning if an entry already exists).  For now, that could just add the other direction as well (also via dict lookup) [^1].

While I doubt that is the perfect API, it seem neither terrible nor irreplaceable enough to actually worry too much that we'll regret adding it.  -- Of course if we e.g. move `bfloat16` into NumPy, then that will subtly change the result for a `from_dlpack()` and import order could also with a warning... But there seems little to be done about that.
In a future where dtypes might want to support `lanes` in a very parametric way things could become a bit less ideal, but it still wouldn't be a cruft that creeps into different parts of NumPy architecture...


[^1]: For the numpy dtype -> dlpack dtype direction I might prefer a method on the DType. But in practice the above is simpler and covers all we need and it doesn't _prevent_ us from adding that method in the future if there is need (e.g. because someone actually wants to support "lanes" or crazy things like that).

**Comment by mattip** [comment]
Can we add the dlpack enum as a property of dtypes, and any dtype without a valid value will raise? Then ml_dtypes.float16 will need to add the property and it will JustWork.

**Comment by seberg** [comment]
> Can we add the dlpack enum as a property of dtypes, and any dtype without a valid value will raise?

If all we have is the DLPack ID on the DType class, we have to assume no-parametrization. I.e. that is byte-order (numpy side) or lanes (DLPack side) parameters don't matter.
But we can't tag it on the dtype instance (it means expanding the ABI and would be painful both in NumPy and for downstream).

So in principle I prefer tagging it on to the DType class! But it may need methods for both directions of conversions to deal with parametrization. And while for `DType -> DLpack` I am very sure that is the nice way to do it, for the opposite direction I dunno if "DLPack ID + method" isn't a bit much...
Because of that, I guess the single (two-way) mapping based on instances seems like it might be a good enough solution to me.  Even if a more thorough solution may eventually be needed (e.g. allowing to map not a dtype instance but a DType class that can deal with parametrization).

The second reason is, that tagging it on to the DType is possibly less convenient to allow `ml_dtypes` quick adoption.  (Although, I am thinking we should allow calling the new-style DType initialization on old-style DTypes to add features that the old style didn't support. I.e. a way to more easily transition over, and allowing to phase out almost all old-style API without a hard break.)

**Comment by mattip** [comment]
> If all we have is the DLPack ID on the DType class, we have to assume no-parametrization. I.e. that is byte-order (numpy side) or lanes (DLPack side) parameters don't matter.

Not clear to me how `register_to_dlpack_map` will handle byte-order and lanes. I proposed putting this all on the dtype instance or class since the information has to live somewhere, and keeping it close to the dtype makes more sense to me than a global registry. For the DLPack -> DType conversion, we could search the registered dtypes for a known DLPackID. 

> tagging it on to the DType is possibly less convenient to allow ml_dtypes quick adoption

Well, they would have to call `register_to_dlpack_map` so some change is required.

**Comment by seberg** [comment]
> Not clear to me how register_to_dlpack_map will handle byte-order and lanes. 

Right, it would side-step the byte-order issue while lanes are pretty irrelevant for now.

I would _prefer_ it defined on the DType and for the `DType->DLPack` direction it is even rather obvious.
But, I suppose I don't have nice plan for that yet. And a Python level call like that works with instances doesn't seem bad.  I.e. simple to explain, doesn't seem like a terrible dead-end API/implementation wise, and good enough for any current foreseeable need.

**Comment by hpkfft** [comment]
If you're considering a more general registration mechanism, it may be worth thinking about the buffer protocol as well.
For example, see also #31204 and #14753

**Comment by hpkfft** [comment]
I see that `ml_dtypes` has `bcomplex32` for `std::complex<bfloat16>`.

Although, I don't know what to do for DLPack, which doesn't seem to have an enumeration for it.
I suppose NumPy could invent something, or even repurpose `bits` or `lanes`.  It seems `kDLBfloat` is only for bfloat16, since bfloat8 gets its own enumeration `kDLFloat8_e5m2`.  So, maybe set `bits = 32` to mean `bcomplex32`?  Or, set `lanes = 2` ?

**Comment by rgommers** [comment]
> Although, I don't know what to do for DLPack, which doesn't seem to have an enumeration for it.

I had a look at the history: that needs adding indeed. I'll open a new issue for that on the DLPack repo (EDIT: see https://github.com/dmlc/dlpack/issues/184)
